### PR TITLE
Initial validation logic for registration fields

### PR DESCRIPTION
**Pull request by jryans**
**Reviewer:** Each commit in this PR is self-contained and has some additional context in the commit message, so you may want to review commit by commit. This PR currently leaves the registration form in a half-converted state with two validation systems. If the code generally looks good, but you'd prefer that we only merge the fully converted version, we can do that instead. Mainly, I want to make sure this approach seems broadly reasonable.

This adds a general validation helper and wires it up to the username field on registration as an example. Validation is handled by describing a rule to check and text to show for each state, which is then shown in a tooltip.

Various TODOs are sprinkled in to mark bits of the previous validation that can be removed after the next phase of converting the other validation checks.

auth-validation-first-cut

Part of the work for https://github.com/vector-im/riot-web/issues/8170

### CODE REVIEW

**Comment by bwindels** [comment]
>  If the code generally looks good, but you'd prefer that we only merge the fully converted version, we can do that instead.

given we want to do an RC as soon as we can, that might be good. Or at least only merged after we create the RC.

File: src/components/views/elements/Validation.js [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        let details;
+        if (results && results.length) {
+            details = <ul className="mx_Validation_details">
+                {results.map(result => {
+                    const classes = classNames({
+                        "mx_Validation_detail": true,
+                        "mx_Validation_valid": result.valid,
+                        "mx_Validation_invalid": !result.valid,
+                    });
+                    return <li key={result.key} className={classes}>
```
Comment (bwindels) [inline src/components/views/elements/Validation.js]: nit: a bit weird that markup/components will get stored in the the state.feedback of the field. I'd rather expect a description object that Field or a subcomponent renders. Not new to this PR though.

Comment (jryans) [reply src/components/views/elements/Validation.js]: Right, we are doing a bit of an unusual thing with design of `Field` and the validation API. I've been quite unsure about it so far, and prototyped quite a few other approaches, such as having  a separate component that contains the `Field`, but none of them have really felt great, so I've just been going with this to get something in place.

**Comment by jryans** [comment]
> > If the code generally looks good, but you'd prefer that we only merge the fully converted version, we can do that instead.
> 
> given we want to do an RC as soon as we can, that might be good. Or at least only merged after we create the RC.

Thanks for the review! I have pushed the reviewed parts to the [jryans/auth-validation branch](https://github.com/matrix-org/matrix-react-sdk/commits/jryans/auth-validation) and will use that as the base for the next phase, then merge it all to develop when ready.

**Pull request closed without merging**